File: sample/src/main/java/tech/thdev/androidrecyclerview/data/source/image/ImagesMetaLocalRepository.java
package tech.thdev.androidrecyclerview.data.source.image;

import java.util.List;

import io.reactivex.Observable;
import tech.thdev.androidrecyclerview.data.LocalImage;

/**
 * Created by Tae-hwan on 18/10/2016.
 */

public class ImagesMetaLocalRepository implements ImagesMetaDataSource {

    private ImagesMetaLocalDataSource dataSource;

    private ImagesMetaLocalRepository() {
        dataSource = new ImagesMetaLocalDataSource();
    }

    private static ImagesMetaLocalRepository repository;
    public static ImagesMetaLocalRepository getInstance() {
        if (repository == null) {
            repository = new ImagesMetaLocalRepository();
        }
        return repository;
    }

    @Override
    public Observable<List<LocalImage>> getAllImages() {
        return dataSource.getAllImages();
    }
}

File: sample/src/main/java/tech/thdev/androidrecyclerview/data/source/image/ImagesMetaLocalRepository.kt
package tech.thdev.androidrecyclerview.data.source.image

import io.reactivex.rxjava3.core.Flowable
import tech.thdev.androidrecyclerview.data.LocalImage

/**
 * Created by Tae-hwan on 18/10/2016.
 */
class ImagesMetaLocalRepository private constructor() : ImagesMetaDataSource {

    companion object {
        private var repository: ImagesMetaLocalRepository? = null

        fun newInstance(): ImagesMetaLocalRepository =
            repository ?: ImagesMetaLocalRepository().also { repository = it }
    }

    private val dataSource: ImagesMetaLocalDataSource = ImagesMetaLocalDataSource()

    override fun getAllImages(): Flowable<List<LocalImage>> {
        return dataSource.getAllImages()
    }
}